rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+easier to reason about each of the sub-graphs and the overall node graph is
+still strongly constrained.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will be the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Node and Channel Creation
+
+Creation of static nodes and channels is straight-forward. The planner node will
+read these on start-up and use the values specified in the manifest to create
+the nodes and channels and send the read- and write handles to the appropriate
+nodes.
+
+Dynamic node creation will be supported by parameterised sub-graph templates.
+The template will specify a label pattern to match. The pattern will include
+information on extracting the parameter values from the incoming labels. The
+trigger for dynamic node creation will be incoming invocations. The label of the
+channel associated with the request receiver on the invocation will be used to
+match the template label pattern and extract the parameters.
+
+A sub-graph template will also be able to define whether it is persistent or
+not. If it is persistent, an instance will be created the first time it is
+needed for a specific label. For future instances of the same label it will
+reuse the already-created nodes and channels. If it is not persistent, the
+subgraph will be recreated every time a matching label arrives. It is the
+responsibility of the nodes in question to close the channels and terminate once
+processing is complete to avoid running out of resources.
+
+The template will contain definitions of nodes and channels to be created. These
+definitions can use the parameters for specifying labels, names etc. The
+template will support a formatting mechanism to replace placeholders with
+parameter values. In most cases the parameter values will be byte arrays. The
+formatting will allow specification of conversions to strings so that it could
+be used in names etc. Initially only base64 encoding will be supported for
+converting from bytes to strings.
+
+As an example, if a parameter named `user_id` is extracted for the label, a
+user-specific node can use the base64 encoding of the user id to create a unique
+name:
+
+```toml
+[node_template]
+name = "user_{base64($user_id)}"
+
+```
+
+The initial impementation of dynamic node creation will focus only on support
+for per-user labels, as it is the first motivating example we have. Each
```

Comment:
I'm not sure I understand why the templating logic even needs to know about the difference between labels, could you elaborate? I was assuming it would just define a number of "free" (i.e. unbound) label parameters, and bind them with whatever concrete label happens to be passed in, but it does not need to know anything about the structure. Going with the generic type analogy, a generic type parameter is, well, generic (except when it needs to specify a type bound, but that does not seem to be the case here).

Reply:
Assuming the planner node is the initial node that is created and it is a reusable node with a fixed hash: it needs to know which template to select for an incoming gRPC invocation it is receiving. If all templates just have unspecified values it does not know which one to choose. The idea is that the label template can have some concrete tags (e.g a Wasm hash tag) and some parameterised template tags (e.g. a user tag). If the planner node sees an incoming label with a wasm tag that matches the concrete tag in the template and a user tag, it will match the template's pattern and extract the value of the user tag as the parameter value to use in the rest of the template. At some point in a sub-graph a node could declassify the wasm tag away, so just the user tag will remain in the label.

You can think of the concrete tags in the label template type constraint on generic type parameters.